r/R/dplyr-funcs-datetime.R

```diff
@@ -148,4 +148,56 @@ register_bindings_datetime <- function() {
     !call_binding("am", x)
   })
 
+  register_binding("ymd", function(x) {
+    format_map <-
+      list(
+        ymd_hyphen1 = "%Y-%m-%d",
+        ymd_hyphen2 = "%y-%m-%d",
+        ymd_hyphen3 = "%Y-%B-%d",
+        ymd_hyphen4 = "%y-%B-%d",
+        ymd_hyphen5 = "%Y-%b-%d",
+        ymd_hyphen6 = "%y-%b-%d",
```

Comment:
Could we use one set of 6 like this, and then pre-process the strings with a regex like was suggested in https://issues.apache.org/jira/browse/ARROW-14471?focusedCommentId=17446011&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17446011 ?

Something kind of like:

```
format_map <- list(
        ymd_hyphen1 = "%Y-%m-%d",
        ymd_hyphen2 = "%y-%m-%d",
        ymd_hyphen3 = "%Y-%B-%d",
        ymd_hyphen4 = "%y-%B-%d",
        ymd_hyphen5 = "%Y-%b-%d",
        ymd_hyphen6 = "%y-%b-%d"
)

x <- gsub("[^A-Za-z0-9.]", "-", x)
call_binding(
      "coalesce",
      call_binding("strptime", x, format = format_map[[1]], unit = "s"),
      call_binding("strptime", x, format = format_map[[2]], unit = "s"),
      call_binding("strptime", x, format = format_map[[3]], unit = "s"),
      call_binding("strptime", x, format = format_map[[4]], unit = "s"),
      call_binding("strptime", x, format = format_map[[5]], unit = "s"),
      call_binding("strptime", x, format = format_map[[6]], unit = "s")
    )
```

You might need to use `call_binding("gsub", ...)` instead of being able to use it directly

Reply:
This would at least cut down on the number of formats we need to try and _mostly_ get the right answers

Reply:
I think I will move the focus on `parse_date_time()` of which most other date/time parsing functions are a particular case of. `ymd(x)` becomes `parse_date_time(x, orders = "ymd")`, but the problems highlighted here still remain.

Reply:
>  coalesce() needs the first call to call_binding("strptime") to return NA in order to move on to the second one and so on. Instead the first call errors and the second one never get evaluated. tryCatch() doesn't seem to work in this context.

Good point! Is there a jira to make that a possibility for the `strptime` binding (or in the C++ itself, which is the ultimate place it'll really need to be...)?

Reply:
https://issues.apache.org/jira/browse/ARROW-15659, but this is for the R side only.